File: EPIJudge/epi_judge_cpp/buy_and_sell_stock.cc
#include <vector>
#include "test_framework/generic_test.h"

using std::vector;

double BuyAndSellStockOnce(const vector<double>& prices) {
	if (prices.size() < 2) {
		return 0;
	}

	double smallest = prices[0];
	double maxProfit = prices[1] - prices[0];

	for (int i = 1; i < prices.size(); i++) {
		maxProfit = std::max(maxProfit, prices[i] - smallest);
		smallest = std::min(smallest, prices[i]);
	}

	return maxProfit > 0 ? maxProfit : 0;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"prices"};
  return GenericTestMain(args, "buy_and_sell_stock.tsv", &BuyAndSellStockOnce,
                         DefaultComparator{}, param_names);
}

File: EPIJudge/epi_judge_python/buy_and_sell_stock.py
def buy_and_sell_stock_once(prices):
	if len(prices) < 2:
		return 0
		
	smallest = prices[0]
	maxProfit = -float('inf')

	for i in range(1, len(prices)):
		maxProfit = max(maxProfit, prices[i] - smallest)
		smallest = min(smallest, prices[i])

	return maxProfit if maxProfit > 0 else 0


from sys import exit

from test_framework import generic_test, test_utils

if __name__ == '__main__':
    exit(
        generic_test.generic_test_main("buy_and_sell_stock.tsv",
                                       buy_and_sell_stock_once))
